### ISSUE TITLE
Seed value of 0 produces inconsistent results

### ISSUE THREAD

**Issue by kennethdmiller3** [lmccart/p5.js#469]
A seed value of 0 is treated as false so both instances of lcg.setSeed fall through to the Math.round(Math.random() * m) case. As a result, you get a different randomly-selected seed each time instead of the seed you asked for.

Here's an alternative implementation that permits a seed of 0 and ensures that the seed is an integer between 0 and m-1:
```javascript
setSeed: function (val) {
  z = seed = Math.floor((val === undefined) ? Math.random() * m : val);
}
```

**Comment by GoToLoop** [comment]
```JavaScript
setSeed : function(val) {
  z = seed = val == null? Math.random()*m | 0 : val | 0;
}
```

I believe the style above is more flexible, since it checks against both `undefined` & `null` and coerces *val* to be a whole number!  :D

**Comment by kennethdmiller3** [comment]
Nice. I'm coming from a C/C++ programming background so the Javascript programming idioms are new to me.

**Comment by GoToLoop** [comment]
I have a little C background too! And Java as well! JS bitshift operators coerces everything to whole values:  

```JS
var arr = [Math.PI];
console.log(arr); // prints out 3!
```

**Comment by kennethdmiller3** [comment]
Based on what I've read, the "| 0" idiom is faster than Math.floor. It truncates, though, so it's not a direct replacement. We're dealing with non-negative values here so it doesn't matter so much.

**Comment by GoToLoop** [comment]
Since both answers needs to be truncated, this 1 is even better, like you did:  
z = seed = val == (null? Math.random()*m : val)  | 0;

**Comment by kennethdmiller3** [comment]
I deleted and re-forked my copy of the repository and will try again using a branch so my unrelated changes don't get mixed in.

**Comment by kennethdmiller3** [comment]
The absolute value thing isn't necessary, I don't think.

**Comment by GoToLoop** [comment]
It's just a safe-guard against negative value arguments! **Math.abs()** was added to **ellipse()** so it accepts negative dimensions!  o_0

**Comment by kennethdmiller3** [comment]
The ideal behavior would be to wrap around like an unsigned int.

**Comment by GoToLoop** [comment]
JS only got `number` datatype to deal w/ number values. It's similar to `signed double` I guess.

**Comment by kennethdmiller3** [comment]
I figured it out: mask the value with an "and". There's a couple ways of doing that::
```javascript
z = seed = (val == null ? Math.random() * m : val ) & 0xFFFFFFFF;
z = seed = (val == null ? Math.random() * m : val ) & (m - 1);
```

**Comment by GoToLoop** [comment]
Bitwise operator `>>>` yields crazy values for negative numbers:  O_o
`console.log(-10.5 >>> 0);  // 4294967286`

**Comment by kennethdmiller3** [comment]
The p5.js linear congruential generator uses unsigned 32-bit math so that's the expected (and desired?) behavior.

Java's 48-bit (!) linear congruential generator uses 64-bit math. I managed to replicate it in Javascript but the resulting generator is complicated and expensive since it involves breaking the 48-bit operations into six 16x16->32 bit multiplies. The division and modulo operations don't help either.

The p5.js generator could use ">>> 0" instead of "% m" to perform the 32-bit modulo, though I decided to avoid messing with that for now.

**Comment by kennethdmiller3** [comment]
Since the generator uses a modulo of 2^32, rand() could use the same ">>> 0" trick to produce the same sequence but slightly faster:
```javascript
z = (a * z + c) >>> 0
return z / m
```
This trick only works for an m of 2^32, of course, though that's why one would use an m of 2^32 instead of some large prime number.

I wonder if inlining the modulo value would improve performance any:
```javascript
z = seed = (val == null ? Math.random() * 0x100000000 : val) >>> 0
...
z = (a * z + c) >>> 0
return z / 0x100000000
```

**Comment by kennethdmiller3** [comment]
As an aside, the "a" and "b" values are about as large as they can be without losing precision during multiplication. Evaluating the formula with the maximum possible z value (1664525 * 5294967295 + 1013904223) yields a result (7149081450614098) with 53 bits, which uses every last bit of precision in a double-precision floating-point value.

(I confirmed that it's working like it should by comparing the result from the Firefox Javascript console as I got from a calculator. I got the same result in both cases.)
